rust/src/io/commit/dynamodb.rs

```diff
@@ -0,0 +1,494 @@
+    () => {
+        "base_uri"
+    };
+}
+
+macro_rules! version {
+    () => {
+        "version"
+    };
+}
+
+static BASE_URI: &str = base_uri!();
+static VERSION: &str = version!();
+static PATH: &str = "path";
+
+impl DynamoDBExternalManifestStore {
+    #[allow(dead_code)]
+    pub async fn new_external_store(
+        client: Client,
+        table_name: String,
+    ) -> Result<Arc<dyn ExternalManifestStore>> {
+        // WHY ARE AWS TYPES DESIGNED SO POORLY -- WHY IS EVERYTHING OPTION<T>? EVEN WHEN VALUES SHOULD NEVER BE NONE
+
+        // Check if the table schema is correct
+        let describe_result = client
+            .describe_table()
+            .table_name(&table_name)
+            .send()
+            .await?;
+        let table = describe_result.table.context(IOSnafu {
+            message: format!("dynamodb table: {table_name} does not exist"),
+        })?;
+        let mut schema = table.key_schema.context(IOSnafu {
+            message: format!("dynamodb table: {table_name} does not have a key schema"),
+        })?;
+
+        let mut has_hask_key = false;
+        let mut has_range_key = false;
+
+        // there should be two keys, HASH(base_uri) and RANGE(version)
+        for _ in 0..2 {
+            let key = schema.pop().context(IOSnafu {
+                message: format!("dynamodb table: {table_name} must have HASH and RANGE keys"),
+            })?;
+            let key_type = key.key_type.context(IOSnafu {
+                message: format!("dynamodb table: {table_name} key types must be defined"),
+            })?;
+            let name = key.attribute_name.context(IOSnafu {
+                message: format!("dynamodb table: {table_name} key must have an attribute name"),
+            })?;
+            match (key_type, name.as_str()) {
+                (KeyType::Hash, base_uri!()) => {
+                    has_hask_key = true;
+                }
+                (KeyType::Range, version!()) => {
+                    has_range_key = true;
+                }
+                // This should never happen but aws generated the branch anyways
+                _ => {
+                    return Err(Error::IO {
+                        message: format!(
+                            "dynamodb table: {table_name} unknown key type encountered name:{name}"
+                        ),
+                    });
+                }
+            }
+        }
+
+        // Both keys must be present
+        if !(has_hask_key && has_range_key) {
+            return Err(
+                Error::IO {
+                    message: format!("dynamodb table: {table_name} must have HASH and RANGE keys, named `{BASE_URI}` and `{VERSION}` respectively")
+                }
+            );
+        }
+
+        Ok(Arc::new(Self { client, table_name }))
+    }
+
+    fn put(&self) -> PutItemFluentBuilder {
+        self.client.put_item().table_name(&self.table_name)
+    }
+
+    fn get(&self) -> GetItemFluentBuilder {
```

Comment:
This share the same name as a public API `ExtenralManifestStore::get()` with different signature, shall we rename this function?

Reply:
I'll rename these functions with `ddb_` prefix